rclcpp/test/test_node.cpp

```diff
@@ -1360,26 +1362,339 @@ TEST_F(TestNode, get_parameters_undeclared_parameters_allowed) {
+      EXPECT_EQ(result.name, name1);
+      EXPECT_EQ(result.type, rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+      EXPECT_FALSE(result.read_only);
+    }
+    {
+      auto result = node->describe_parameter(name2);
+      EXPECT_EQ(result.name, name2);
+      EXPECT_EQ(result.type, rcl_interfaces::msg::ParameterType::PARAMETER_STRING);
+      EXPECT_TRUE(result.read_only);
+    }
+  }
+  {
+    // non-existent parameter does not throw, but returns default constructed one
+    auto name = "parameter"_unq;
+
+    {
+      auto result = node->describe_parameter(name);
+      rcl_interfaces::msg::ParameterDescriptor expected;
+      expected.name = name;
+      EXPECT_EQ(result, expected);
+    }
+  }
+}
+
+// test describe parameters with undeclared not allowed
+TEST_F(TestNode, describe_parameters_undeclared_parameters_not_allowed) {
+  auto node = std::make_shared<rclcpp::Node>(
+    "test_get_parameters_node"_unq,
+    rclcpp::NodeOptions().allow_undeclared_parameters(false));
+  {
+    // normal use
+    auto name1 = "parameter"_unq;
+    auto name2 = "parameter"_unq;
+
+    node->declare_parameter(name1, 42);
+    rcl_interfaces::msg::ParameterDescriptor descriptor;
+    descriptor.read_only = true;
+    node->declare_parameter<std::string>(name2, "test", descriptor);
+
+    auto results = node->describe_parameters({name1, name2});
+
+    EXPECT_EQ(results.size(), 2u);
+
+    EXPECT_EQ(results[0].name, name1);
+    EXPECT_EQ(results[0].type, rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+    EXPECT_FALSE(results[0].read_only);
+
+    EXPECT_EQ(results[1].name, name2);
+    EXPECT_EQ(results[1].type, rcl_interfaces::msg::ParameterType::PARAMETER_STRING);
+    EXPECT_TRUE(results[1].read_only);
+  }
+  {
+    // non-existent parameter throws
+    auto name = "parameter"_unq;
+
+    {
+      EXPECT_THROW({
+        node->describe_parameters({name});
+      }, rclcpp::exceptions::ParameterNotDeclaredException);
+    }
+  }
+  {
+    // check that repeated names in input work, and that output is stable (same order as input)
+    auto name1 = "parameter"_unq;
+    auto name2 = "parameter"_unq;
+
+    node->declare_parameter(name1, 42);
+    rcl_interfaces::msg::ParameterDescriptor descriptor;
+    descriptor.read_only = true;
+    node->declare_parameter<std::string>(name2, "test", descriptor);
+
+    auto results = node->describe_parameters({name2, name1, name2});
+
+    EXPECT_EQ(results.size(), 3u);
+
+    EXPECT_EQ(results[0].name, name2);
+    EXPECT_EQ(results[0].type, rcl_interfaces::msg::ParameterType::PARAMETER_STRING);
+    EXPECT_TRUE(results[0].read_only);
+
+    EXPECT_EQ(results[1].name, name1);
+    EXPECT_EQ(results[1].type, rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+    EXPECT_FALSE(results[1].read_only);
+
+    EXPECT_EQ(results[2].name, name2);
+    EXPECT_EQ(results[2].type, rcl_interfaces::msg::ParameterType::PARAMETER_STRING);
+    EXPECT_TRUE(results[2].read_only);
+  }
```

Comment:
Consider also testing with a declared parameter and undeclared parameter together.